storage/allocator.go

```diff
@@ -13,75 +13,291 @@
+// to store replicas.
+func getUsedNodes(existing []proto.Replica) map[proto.NodeID]struct{} {
 	// Get a set of current nodes -- we never want to allocate on an existing node.
-	usedNodes := make(map[proto.NodeID]struct{})
-	for _, replica := range existingReplicas {
+	usedNodes := map[proto.NodeID]struct{}{}
+	for _, replica := range existing {
 		usedNodes[replica.NodeID] = struct{}{}
 	}
+	return usedNodes
+}
+
+// storeDescFromGossip retrieves a StoreDescriptor from the specified
+// capacity gossip key. Returns an error if the gossip doesn't exist
+// or is not a StoreDescriptor.
+func storeDescFromGossip(key string, g *gossip.Gossip) (*proto.StoreDescriptor, error) {
+	info, err := g.GetInfo(key)
 
-	stores, err := a.storeFinder(required)
 	if err != nil {
 		return nil, err
 	}
+	storeDesc, ok := info.(proto.StoreDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("gossiped info is not a StoreDescriptor: %+v", info)
+	}
+	return &storeDesc, nil
+}
+
+// capacityGossipUpdate is a gossip callback triggered whenever capacity
+// information is gossiped. It just tracks keys used for capacity
+// gossip.
+func (a *allocator) capacityGossipUpdate(key string, contentsChanged bool) {
+	a.Lock()
+	defer a.Unlock()
+
+	// Clear the cached store lists on new gossip.
+	a.storeLists = nil
+	if a.capacityKeys == nil {
+		a.capacityKeys = map[string]struct{}{}
+	}
+	a.capacityKeys[key] = struct{}{}
+}
 
-	// Randomly pick a node weighted by capacity.
-	var candidates []*proto.StoreDescriptor
-	var capacityTotal float64
+// Allocate returns a suitable store for a new allocation with the
+// required attributes. Nodes already accommodating existing replicas
+// are ruled out as targets.
+func (a *allocator) Allocate(required proto.Attributes, existing []proto.Replica) (
+	*proto.StoreDescriptor, error) {
+	a.Lock()
+	defer a.Unlock()
+	stores, err := a.selectRandom(3, required, existing, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Choose the store with the highest percent capacity available.
+	var highStore *proto.StoreDescriptor
 	for _, s := range stores {
-		if _, ok := usedNodes[s.Node.NodeID]; !ok {
-			candidates = append(candidates, s)
-			capacityTotal += s.Capacity.PercentAvail()
+		if highStore == nil || s.Capacity.PercentAvail() > highStore.Capacity.PercentAvail() {
+			highStore = s
+		}
+	}
+	return highStore, nil
+}
+
+// Rebalance returns a suitable store for a rebalance target with
+// required attributes. Rebalance targets are selected at random
+// from amongst stores which fall within an acceptable multiple
+// of standard deviations from either the capacity or range count
+// means.
+func (a *allocator) Rebalance(required proto.Attributes, existing []proto.Replica) (
+	*proto.StoreDescriptor, error) {
+	a.Lock()
+	defer a.Unlock()
+	filter := func(s *proto.StoreDescriptor, count, capacity *stat) bool {
+		// Use counts instead of capacities if the cluster has mean capacity
```

Comment:
I don't think a fuzzy threshold (presumably with different probabilities of selection) would add very much utility. It would add complexity. Perhaps you had something else in mind?